internal/core/src/index/HybridScalarIndex.cpp

```diff
@@ -0,0 +1,388 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "index/HybridScalarIndex.h"
+#include "common/Slice.h"
+#include "common/Common.h"
+#include "index/Meta.h"
+#include "index/ScalarIndex.h"
+#include "index/Utils.h"
+#include "storage/Util.h"
+#include "storage/space.h"
+
+namespace milvus {
+namespace index {
+
+template <typename T>
+HybridScalarIndex<T>::HybridScalarIndex(
+    const storage::FileManagerContext& file_manager_context)
+    : is_built_(false) {
+    if (file_manager_context.Valid()) {
+        file_manager_ =
+            std::make_shared<storage::MemFileManagerImpl>(file_manager_context);
+        AssertInfo(file_manager_ != nullptr, "create file manager failed!");
+    }
+    internal_index_type_ = InternalIndexType::NONE;
+}
+
+template <typename T>
+HybridScalarIndex<T>::HybridScalarIndex(
+    const storage::FileManagerContext& file_manager_context,
+    std::shared_ptr<milvus_storage::Space> space)
+    : is_built_(false), space_(space) {
+    if (file_manager_context.Valid()) {
+        file_manager_ = std::make_shared<storage::MemFileManagerImpl>(
+            file_manager_context, space);
+        AssertInfo(file_manager_ != nullptr, "create file manager failed!");
+    }
+    internal_index_type_ = InternalIndexType::NONE;
+}
+
+template <typename T>
+InternalIndexType
+HybridScalarIndex<T>::SelectIndexBuildType(size_t n, const T* values) {
+    std::set<T> distinct_vals;
+    for (size_t i = 0; i < n; i++) {
+        distinct_vals.insert(values[i]);
```

Comment:
break when distinct_vals.size() >= BITMAP_INDEX_CARDINALITY_BOUND

Reply:
this interface is just used for test, test data common not big, not a serious problem. the real work interface do it